### PR TITLE
WalletConnect Integration

### PR DESCRIPTION
**Pull request by blurpesec**
### Enables WalletConnect on MyCrypto Beta

ToDo:
- [x] Add-Account Flow
- [x] Flip the name back over 
- [x] Sign Tx Flow
- - [x] Detection of WalletConnect
- - [x] SignTx Panel (_needs some design suggestions for dis_)
- [x] Delete standalone `/common/v2/components/WalletConnectQr.tsx` when we're sure it's no longer needed
- [x] Issue with re-mounting qr causing session to reset.

References:
1) https://example.walletconnect.org/
2) https://docs.walletconnect.org/quick-start/dapps
3) https://docs.walletconnect.org/quick-start/wallets (relevant for Ambo)

### Views:
Sign Tx w/WalletConnect - Error - resubmitting
image

Add-Account WalletConnect
image

### CODE REVIEW

File: common/v2/components/SignTransactionWallets/WalletConnect.tsx [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+enum WalletSigningState {
+  READY, // when signerWallet is ready to sendTransaction
+  UNKNOWN // used upon component initialization when wallet status is not determined
+}
+
+export type WalletConnectQrContent = WalletConnectAddress;
+
+//const WalletService = WalletFactory(WalletId.WALLETCONNECT);
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply common/v2/components/SignTransactionWallets/WalletConnect.tsx]: done!

File: common/v2/components/SignTransactionWallets/WalletConnect.tsx [hunk 2]
```diff
@@ -0,0 +1,175 @@
+import React, { useState, useEffect } from 'react';
+import translate, { translateRaw } from 'v2/translations';
+
+import { WalletConnectSignQr, Button, Spinner } from 'v2/components';
+//import { WalletFactory } from 'v2/services/WalletService';
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply common/v2/components/SignTransactionWallets/WalletConnect.tsx]: done!

File: common/v2/components/SignTransactionWallets/WalletConnect.reducer.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import { ExtendedAccount, ITxObject } from 'v2/types';
```
Comment (vilsbole) [inline common/v2/components/SignTransactionWallets/WalletConnect.reducer.ts]: Beautiful!

File: common/v2/services/WalletService/walletconnect/walletConnectSingleton.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    // If there is no current active bridge, create one
+    if (!this.instance.bridge) {
+      this.instance = new WalletConnectInstance({
+        bridge: walletConnectBridgeURI
+      });
+    }
+    // If there is no current active session, create one
+    if (this.instance.key === '' && this.instance.bridge) {
+      await this.instance.createSession();
+      return this.instance;
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply common/v2/services/WalletService/walletconnect/walletConnectSingleton.ts]: Nice catch!

File: common/v2/services/WalletService/walletconnect/reducer.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      };
+    }
+    case WalletConnectServiceActions.REFRESH_SESSION: {
+      console.debug('[walletConnectReducer]: querying REFRESH_SESSION');
+      const { session } = payload;
+      return {
+        ...state,
+        session
+      };
+    }
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply common/v2/services/WalletService/walletconnect/reducer.ts]: done in https://github.com/MyCryptoHQ/MyCrypto/pull/2911/commits/07b3a7a35a123fb3c3ec6f1d562a0771fd5680d0

File: common/v2/services/WalletService/walletconnect/WalletConnectProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+export const initWalletConnectState: WalletConnectState = {
+  session: undefined
+};
+
+export const WalletConnectContext = createContext({} as WalletConnectProviderState);
+
+export const WalletConnectProvider = ({ children }: any) => {
+  const [state, dispatch] = useReducer(walletConnectSessionReducer, initWalletConnectState);
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply common/v2/services/WalletService/walletconnect/WalletConnectProvider.tsx]: done in: https://github.com/MyCryptoHQ/MyCrypto/pull/2911/commits/07b3a7a35a123fb3c3ec6f1d562a0771fd5680d0

File: common/v2/services/WalletService/walletconnect/index.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export * from './walletConnectSingleton';
+export * from './WalletConnectProvider';
+export * from './types';
+export * from './reducer';
```
[the remark this answers is not in the document]
Comment (blurpesec) [reply common/v2/services/WalletService/walletconnect/index.ts]: > Prefer explicit exports for a first level module.

What is considered a first-level module?

**Comment by 3esmit** [comment]
Can MyCrypto be also a WalletConnect provider? Like, you open the key on MyCrypto and then it can connect in WalletConnect on other Dapp?

**Comment by blurpesec** [comment]
> Can MyCrypto be also a WalletConnect provider? Like, you open the key on MyCrypto and then it can connect in WalletConnect on other Dapp?

Its not currently planned, since that would be a bit of a fundamental change to the way the application works. Right now private keys/json/keystore/etc things only exist for the brief period of time needed to add the public address they map to, to the local storage, and the brief period of time where they're used to sign transactions.

This feature would be hard to use because walletconnect wallet-side signing is primarily meant for a mobile phone's mobility. i.e -> It'd be hard to scan a qr code on web-based dapp with a scanner also based on the web.

**Comment by vilsbole** [comment]
Closing in favour of #3024

**Pull request closed without merging**